sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
     bufferMap.compute(buffer.id, updater)
   }
 
-  /** Remove a buffer ID from the catalog at the specified storage tier. */
-  def removeBufferTier(id: RapidsBufferId, tier: StorageTier): Unit = {
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  def synchronousSpill(store: RapidsBufferStore, targetTotalSize: Long): Option[Long] =
+    synchronousSpill(store, targetTotalSize, Cuda.DEFAULT_STREAM)
+
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param store store to spill from
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @param stream CUDA stream to use or null for default stream
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  private def synchronousSpill(
+      store: RapidsBufferStore,
+      targetTotalSize: Long,
+      stream: Cuda.Stream): Option[Long] = {
+    require(targetTotalSize >= 0, s"Negative spill target size: $targetTotalSize")
+    logWarning(s"Targeting a ${store.name} size of $targetTotalSize. " +
+      s"Current total ${store.currentSize}. " +
+      s"Current spillable ${store.currentSpillableSize}")
+
+    // we try to spill in this thread. If another thread is also spilling, we let that
+    // thread win and we return letting RMM retry the alloc
+    var rmmShouldRetryAlloc = false
+
+    // total amount spilled in this invocation
+    var totalSpilled: Long = 0
+
+    if (store.currentSpillableSize > targetTotalSize) {
+      withResource(new NvtxRange(s"${store.name} sync spill", NvtxColor.ORANGE)) { _ =>
+        logWarning(s"${store.name} store spilling to reduce usage from " +
+          s"${store.currentSize} total (${store.currentSpillableSize} spillable) " +
+          s"to $targetTotalSize bytes")
+
+        // waited: the store spilled, but it didn't free right away. Instead it waits for
+        // buffers to be closed by clients (since they had references)
+        var waited = false
+
+        // If the store has 0 spillable bytes left, it has exhausted.
+        var exhausted = false
+
+        while (!exhausted && !rmmShouldRetryAlloc &&
+            store.currentSpillableSize > targetTotalSize) {
+          val mySpillCount = spillCount
+          var spilled = false
```

Comment:
I think I partially agree. The main difference is that with `totalSpilled > 0` we will not wait for pending bytes, because once the first buffer is spilled this total stays above 0. We also won't set `exhausted=true` and will rely on `store.currentSpillableSize` to be <= targeTotalSize to exit the loop. I don't think this is a live lock scenario, but it is different behavior, and I think would make us more aggressive at attempting to realloc.

I would then want to remove the `waitForPending` wait. It seems we would stop needing that.

Reply:
Yeah, per our previous discussions, I think we would want to free what we can without waiting and retry the alloc rather than assume the alloc retry will fail if we freed something but it may not be enough and assume we need to wait for more memory to free before retrying.

Reply:
Ok I'll push a commit that cleans this up then, it should remove a question @revans2 raised on `waitForPending` as well

Reply:
https://github.com/NVIDIA/spark-rapids/pull/7572/commits/bc2a2150bfa8be133a14e9ee5a465eb0fd9d10ab This removes the waiting logic.